File: Day72/Part1/random_seed.py
def solve():
    try:
        line = input().split()
        if not line:
            return
        # Check if first token is n or array
        # Assuming standard format: n then array, or just array
        # If line has 1 element, likely n.
        if len(line) == 1:
            n = int(line[0])
            line2 = input().split()
            seeds = [int(x) for x in line2]
        else:
            seeds = [int(x) for x in line]
            
        if not seeds:
            print(0)
            return
            
        dp = {}
        global_max = 0
        
        for x in seeds:
            current_max = 0
            # Find divisors
            # Optimization: iterate up to sqrt(x)
            i = 1
            while i * i <= x:
                if x % i == 0:
                    d1 = i
                    d2 = x // i
                    
                    if d1 in dp:
                        current_max = max(current_max, dp[d1])
                    if d2 in dp:
                        current_max = max(current_max, dp[d2])
                i += 1
            
            dp[x] = current_max + 1
            global_max = max(global_max, dp[x])
            
        print(global_max)
            
    except EOFError:
        pass

if __name__ == "__main__":
    solve()

File: Day72/Part1/random_seed.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

int main() {
    int n;
    if (cin >> n) {
        vector<int> seeds(n);
        for (int i = 0; i < n; ++i) {
            cin >> seeds[i];
        }

        if (n == 0) {
            cout << 0 << endl;
            return 0;
        }

        // Map to store max length ending with value v
        map<int, int> dp;
        int global_max = 0;

        for (int x : seeds) {
            int current_max = 0;
            // Find divisors
            for (int i = 1; i * i <= x; ++i) {
                if (x % i == 0) {
                    // Divisor i
                    if (dp.count(i)) {
                        current_max = max(current_max, dp[i]);
                    }
                    // Divisor x/i
                    int d2 = x / i;
                    if (d2 != i) {
                        if (dp.count(d2)) {
                            current_max = max(current_max, dp[d2]);
                        }
                    }
                }
            }
            dp[x] = current_max + 1;
            global_max = max(global_max, dp[x]);
        }

        cout << global_max << endl;
    }
    return 0;
}
